### DISCUSSION TITLE
GitHub Actions cant find branch (resolved)

### DISCUSSION THREAD

**Discussion by bazoukiwi** [community/community#31393] [Actions] [answered]
Hello everyone. I have a NestJS app on github that i frequently push. I now want to set up a workflow to automaticaly push my app on my hoster, evenode, everytime i push to origin main. For now, to push to my host, i re stup git (remote add, etc..) and then git push evennode master. 
But if i configure my workflow to push evennode master, it says "error: src refspec master does not match any".
Here is my yml file, directly taken from the evennode doc:
(ref: https://github.com/evennode/github-actions-deploy-to-evennode)

# This workflow will automatically deploy your application to www.evennode.com on push to main branch.

name: Deploy to EvenNode

on:
  push:
    branches: [ main ]

jobs:
  Deploy:
    runs-on: ubuntu-latest

    steps:
    - name: Install SSH key
      uses: shimataro/ssh-key-action@v2
      with:
        key: ${{secrets.SSH_KEY}}
        known_hosts: " "
    - name: Add git.evennode.com to known_hosts
      run: |
        ssh-keyscan -H git.evennode.com > ~/.ssh/known_hosts
    - uses: actions/checkout@v2
      with:
        fetch-depth: 0
    - name: Remote add evenode
      run: |
        git remote add evennode ${{secrets.EVENNODE_REPO_URL}}
    - name: Config
      run: |
        git config --global user.email "${{secrets.GIT_EMAIL}}"
        git config --global user.name "${{secrets.GIT_NAME}}"
    - name: Push
      run: |
        git push evennode master
        
        
I do have set my secrets
Thank you very much for your time

**Comment by airtower-luna** [comment, marked as answer]
Your workflow runs on a push to **main**, but you're trying to push a ref called **master** to evennode. Change the push to push `main` instead:

```
git push evennode main
```

Or, if evennode by some bad design requires the branch to be called `master` on their side, tell Git to push local branch `main` to remote branch `master`:

```
git push evennode main:master
```

**Reply by bazoukiwi** [reply]
Thanks a lot for you precise response.
I re-initialized git using the branch name "main", on wich i succesfully pushed localy.
On my yml, i now have "git push evennode main".
My new error is this one:

Run git push evennode main
Warning: Permanently added the ECDSA host key for IP address 'x.x.x.x' to the list of known hosts.
Load key "/home/runner/.ssh/id_rsa": invalid format
git@git.evennode.com: Permission denied (publickey,gssapi-keyex,gssapi-with-mic).
fatal: Could not read from remote repository.

on my secrets, my key look like this: 
ssh-rsa xxxxxxx/xxxxx/xxxx.......xxx= my@email.x
isnt it correct?
This is my first time really using GH, i'm a bit confused, sorry

**Reply by airtower-luna** [reply]
> on my secrets, my key look like this:
> ssh-rsa xxxxxxx/xxxxx/xxxx.......xxx= [my@email.x](mailto:my@email.x)

That looks like your *public* key. To authenticate with an SSH server the workflow needs the matching *private* key. The public part of the key is what the server needs to verify the client has the private key, and grant it access. I strongly recommend you create a separate key pair only for the GH/evennode connection. It's a general security precaution, same as for passwords, tokens, etc.: If the private key somehow gets into the wrong hand it limits the damage if it can be used for only one thing.

As a side note, this part of the workflow (also related to SSH) looks dangerous:

```yaml
- name: Add git.evennode.com to known_hosts
  run: |
    ssh-keyscan -H git.evennode.com > ~/.ssh/known_hosts
```

It's basically like overriding a certificate warning in your browser and accepting any certificate you receive at that moment. The [`ssh-keyscan` documentation](https://manpages.debian.org/bullseye/openssh-client/ssh-keyscan.1.en.html) warns about it, too:

> If an ssh_known_hosts file is constructed using ssh-keyscan without verifying the keys, users will be vulnerable to [man in the middle](https://manpages.debian.org/bullseye/openssh-client/ssh-keyscan.1.en.html#man) attacks. On the other hand, if the security model allows such a risk, ssh-keyscan can help in the detection of tampered keyfiles or man in the middle attacks which have begun after the ssh_known_hosts file was created.

It's better to retrieve and verify the key manually, and then explicitly add it to the `known_hosts` file before making an SSH connection. The action you use to install the private key seems to have an input for that, but I'm not familiar with that specific action.

> This is my first time really using GH, i'm a bit confused, sorry

These things apply to SSH in general, so you'll be able to use them beyond GH, too. :slightly_smiling_face:

**Reply by bazoukiwi** [reply]
After many readings, i did understood all you said ! 
First, about keyscan, i just copied what the Evennode doc gave :(. I now managed to retrieve the keys manually , will definitively add them.
About private/public key, i understood i need to gave a private key. What i dont understand is wich one. 
When hosting to evennode, they ask for a public key forged on my local machine ( https://www.evennode.com/docs/git-deployment ) . I created it, gave it to evennode, so when i git push (localy), my machine compare my local public and the one evennode has, authorising to push (I Think).
So then, about ssh. Workflow need a private key to connect to ssh to evennode to push the code. isnt it ? So i assume i need the ssh key on the evennode machine. They dont even offer ssh ! No keys too. They do offer ftp service, i tried it, no use exept for the files created by the app. 
Getting lost here, i'm sorry. I'm mad they dont offer a guideline, considering swiching. 
A massive massive thank you for you help again. You helped me understand concepts. This is gold !

**Reply by airtower-luna** [reply]
> About private/public key, i understood i need to gave a private key. What i dont understand is wich one.

In asymmetric cryptography keys come in pairs: public and private key. You can have as many key pairs as you want, though if you have many you'll need to tell SSH which one to use for which server. Obviously I don't know which of your public keys you gave to Evennode, but there must've been one if you can push over SSH. :slightly_smiling_face: 

> I created it, gave it to evennode, so when i git push (localy), my machine compare my local public and the one evennode has, authorising to push (I Think).

That's incorrect. Highly simplified, your local machine creates a signature using the private key, which the server can verify using the public key. This mathematically proves to the server that you *have* the private key, without it ever leaving your machine. That's why public key authentication is so much more secure than passwords or tokens. Basically adding the public key to an SSH server as authorized (no matter if Evennode, GitHub, or elsewhere) means "allow access to whoever has the matching private key".

> So then, about ssh. Workflow need a private key to connect to ssh to evennode to push the code.

Exactly, that's what you'll need to have in a secret. And you need to give Evennode the matching public key.

> So i assume i need the ssh key on the evennode machine. They dont even offer ssh !

The description you linked says they do. Not in the sense of a shell login, but the description describes how to use Git over SSH to push data there. That's what the remote URL starting with `git@git.evennode.com` indicates. So something like `ssh -v git@git.evennode.com` should work to test the connection and get debug info.

**Reply by bazoukiwi** [reply]
Woah, you really are a goldmine. Thanks again for all you shared with me and took time to write, i really appreciate.
I so created a pair, put the private in secret, the public on evennode, as you suggested. 
Pushed... WORKED !
Note, i had to put the entire private file, 
-----BEGIN OPENSSH PRIVATE KEY-----
....[...]...
-----END OPENSSH PRIVATE KEY-----

My app is now fully working.
Thank you so so much, for all your help, have a nice day !
